source/extensions/common/tap/admin.cc

```diff
@@ -110,40 +109,203 @@ void AdminHandler::unregisterConfig(ExtensionConfig& config) {
-      PANIC("not implemented");
+    std::vector<envoy::data::tap::v3::TraceWrapper> buffer = trace_buffer->flush();
+
+    ENVOY_LOG(debug, "admin writing buffered trace list to response");
+
+    // Serialize writes to the stream
+    for (size_t i = 0; i < buffer.size(); i++) {
+      // Close stream on final message
+      attached_request->streamMsg(buffer[i], (i + 1) == buffer.size());
     }
+  });
+}
+
+AdminHandler::AttachedRequest::AttachedRequest(AdminHandler* admin_handler,
+                                               const envoy::admin::v3::TapRequest& tap_request,
+                                               Server::AdminStream* admin_stream)
+    : config_id_(tap_request.config_id()), config_(tap_request.tap_config()),
+      admin_stream_(admin_stream), main_thread_dispatcher_(admin_handler->main_thread_dispatcher_) {
+}
 
-    ENVOY_LOG(debug, "admin writing buffered trace to response");
-    Buffer::OwnedImpl output_buffer{output_string};
-    parent.attached_request_.value().admin_stream_->getDecoderFilterCallbacks().encodeData(
-        output_buffer, false);
+AdminHandler::AttachedRequestBuffered::AttachedRequestBuffered(
+    AdminHandler* admin_handler, const envoy::admin::v3::TapRequest& tap_request,
+    Server::AdminStream* admin_stream)
+    : AttachedRequest(admin_handler, tap_request, admin_stream) {
+  const envoy::config::tap::v3::OutputSink& sink =
+      tap_request.tap_config().output_config().sinks(0);
+
+  const uint64_t max_buffered_traces = sink.buffered_admin().max_traces();
+  const uint64_t timeout_ms = PROTOBUF_GET_MS_OR_DEFAULT(sink.buffered_admin(), timeout, 0);
+  trace_buffer_ = std::make_unique<TraceBuffer>(max_buffered_traces);
+  // Start the countdown if provided an actual timeout
+  if (timeout_ms > 0) {
+    timer_ = dispatcher().createTimer(
+        [this, admin_handler] { this->onTimeout(admin_handler->attached_request_); });
+    timer_->enableTimer(std::chrono::milliseconds(timeout_ms));
+  }
+}
+
+void AdminHandler::AttachedRequestBuffered::onTimeout(
+    const std::weak_ptr<AttachedRequest>& weak_attached_request) {
+  // Flush the buffer regardless of size
+  dispatcher().post([weak_attached_request] {
+    // Take temporary ownership - extend lifetime
+    std::shared_ptr<AttachedRequest> attached_request = weak_attached_request.lock();
+    if (!attached_request) {
+      ENVOY_LOG(debug, "Timer Expiry after admin tap request completion");
+      return; // No attached request, flushed already
+    }
+    TraceBuffer* trace_buffer = attached_request->traceBuffer();
+
+    // if the trace buffer has already been flushed short circuit
+    if (trace_buffer->flushed()) {
+      return;
+    }
+
+    std::vector<envoy::data::tap::v3::TraceWrapper> buffer = trace_buffer->flush();
+
+    ENVOY_LOG(debug, "Timer Expiry, admin flushing buffered traces to response");
+
+    // Serialize writes to the stream
+    for (const auto& el : buffer) {
```

Comment:
maybe a slightly more informative variable than 'e1' -- all the reader knows is this is a TraceBuffer, which is evident from context 5 lines above, but I don't know what concept e1 means.